rm_stm/tests: add a stress test for concurrent eviction / replication / snapshots

bharathv opened the pull request:
Fixes a race condition found by the test

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v24.3.x
- [x] v24.2.x
- [x] v24.1.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bharathv committed:
producer_state_manager: make period configurable for testing

bharathv committed:
rm_stm/tests: add a high concurrency test for producer eviction

bharathv commented:
/dt

vbotbuildovich commented:
<!--UNIT FAILURES-->
<!--f167cada-44af-49a8-a964-1bc6defec070-->
<!--UNIT FAILURES-->
<!--c1c2c52e-c6ff-4419-bc2d-8de9544e2ab3-->
<!--UNIT FAILURES-->
<!--90dd7d4f-9514-46c2-a42a-3a0d711517bd-->
<!--body-->
the below tests from https://buildkite.com/redpanda/redpanda/builds/59473#0193aa6d-f8af-4bc0-a603-7db2d9a9a115 have failed and will be retried
```
idempotency_tests_rpunit
rm_stm_tests_rpunit
tx_compaction_tests_rpunit
```
the below tests from https://buildkite.com/redpanda/redpanda/builds/59473#0193aa6d-f8ae-4038-b87b-224ef7d6db46 have failed and will be retried
```
idempotency_tests_rpunit
rm_stm_tests_rpunit
tx_compaction_tests_rpunit
```
the below tests from https://buildkite.com/redpanda/redpanda/builds/59511#0193adc4-6296-4327-809d-1e6a5461ce19 have failed and will be retried
```
tx_compaction_tests_rpunit
```
<!--/body-->

vbotbuildovich commented:
<!--RETRIED FAILURES-->
<!--8d5005cb-b6aa-4e28-9eb7-d20f0a12a749-->
<!--RETRIED FAILURES-->
<!--fdd28078-dc7f-45ca-ab88-d29edb0de1a7-->
<!--RETRIED FAILURES-->
<!--2b133661-84d8-4dbc-9e88-6417c32cb662-->
<!--RETRIED FAILURES-->
<!--cf58f8fe-58f0-4233-909c-aa2ef36f08d7-->
<!--RETRIED FAILURES-->
<!--47a2a26d-d24f-4e4f-a847-401c48e4c455-->
<!--RETRIED FAILURES-->
<!--034cacc4-6e9a-48e2-b055-e0a33d46f36b-->
<!--RETRIED FAILURES-->
<!--00785e94-e70e-40b4-a4e9-f45090a3bea3-->
<!--RETRIED FAILURES-->
<!--ea6e6f38-1a78-486e-b071-e2568b575a02-->
<!--RETRIED FAILURES-->
<!--5f22b936-2e75-4ba5-bf8d-db641dc16e5c-->
<!--body-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59473#0193aac7-182c-416d-a407-d5d8fed9c070
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59473#0193aaca-b157-478e-a0bd-b8d05af2dda3
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59473#0193aaca-b159-4950-ac3b-3b095f4baaa4
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59473#0193aaca-b158-471d-90f8-0f34944599ff
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59473#0193aaca-b15a-4bfb-9305-6ff8770d93d8
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59511#0193ae0a-e01d-4545-922b-dff5cb98b000
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59527#0193af9c-4b29-4ea2-b32e-8f2f97714564
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59552#0193b1bc-a604-4fcd-8330-d64df11b3856
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59569#0193b2eb-407e-47ed-858f-304176113cb4
<!--/body-->

vbotbuildovich commented:
<!--NEW FAILURES-->
<!--7a46580d-aa93-48e5-bc2f-f9524b567f88-->
<!--NEW FAILURES-->
<!--8e39e19a-33f2-4662-891e-f980af1783d0-->
<!--NEW FAILURES-->
<!--2f94b73f-7f25-4e9e-954a-5c00988128e6-->
<!--body-->
non flaky failures in https://buildkite.com/redpanda/redpanda/builds/59473#0193aac7-182e-448b-a560-b3d3f895a559:
```
"rptest.transactions.transactions_test.TransactionsTest.check_pids_overflow_test"
```
non flaky failures in https://buildkite.com/redpanda/redpanda/builds/59473#0193aaca-b159-4950-ac3b-3b095f4baaa4:
```
"rptest.transactions.transactions_test.TransactionsTest.check_pids_overflow_test"
```
non flaky failures in https://buildkite.com/redpanda/redpanda/builds/59511#0193ae06-2587-416a-94ce-c0a4045431fb:
```
"rptest.tests.e2e_shadow_indexing_test.EndToEndHydrationTimeoutTest.test_hydration_completes_when_consumer_killed"
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59473-->
<!--14840d2e-9b0c-4bad-89d1-81983dba3673-->
<!--b137beec-1291-49c5-a3fc-2525d431b872-->
<!--9d2e6ee1-923a-4999-be05-09bba10213f8-->
<!--e6eccdb0-d192-4c6e-a79f-4ee324e0dafb-->
<!--title-->
## Retry command for [Build#59473](https://buildkite.com/redpanda/redpanda/builds/59473)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/transactions/transactions_test.py::TransactionsTest.check_pids_overflow_test
tests/rptest/tests/idempotency_stress_test.py::IdempotencyStressTest.producer_id_stress_test@{"max_producer_ids":100}
tests/rptest/tests/idempotency_stress_test.py::IdempotencyStressTest.producer_id_stress_test@{"max_producer_ids":3000}
tests/rptest/tests/idempotency_stress_test.py::IdempotencyStressTest.producer_id_stress_test@{"max_producer_ids":1000}
```
<!--/body-->

bharathv commented:
/dt

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59511-->
<!--8dd09669-8665-472e-bc41-a323d7f5c573-->
<!--title-->
## Retry command for [Build#59511](https://buildkite.com/redpanda/redpanda/builds/59511)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndHydrationTimeoutTest.test_hydration_completes_when_consumer_killed
```
<!--/body-->

mmaslankaprv commented on src/v/cluster/tests/rm_stm_tests.cc:
```diff
@@ -931,3 +931,95 @@ FIXTURE_TEST(test_tx_expiration_without_data_batches, rm_stm_test_fixture) {
+ * replication operations and operations that reset the state (snapshots,
+ * partition stop).
+ */
+FIXTURE_TEST(test_concurrent_producer_evictions, rm_stm_test_fixture) {
```
can you enable stress fiber in this test ? From my experience it is great in uncovering all sorts of race conditions

bharathv committed:
rm_stm: fix a race condition between async cleanup and reset

This is a classic iterator invalidation caught by the test added in the
previous commit. Cleanup could race with reset thus invalidating the
iterator used in max_concurrent_for_each().

mmaslankaprv commented on src/v/cluster/rm_stm.cc:
```diff
@@ -114,6 +115,18 @@ rm_stm::rm_stm(
+              if (!ssx::is_shutdown_exception(ex)) {
+                  vlog(
+                    _ctx_log.warn,
+                    "encountered an exception while cleaning producers: ",
```
placeholder is missing

bharathv replied:
yikes, fixed.

dotnwat commented:
/rp-unit-test
ctest_args=-R rm_stm_tests_rpunit --repeat until-fail:100

ztlpn commented on src/v/cluster/rm_stm.cc:
```diff
@@ -148,23 +163,37 @@ rm_stm::maybe_create_producer(model::producer_identity pid) {
 };
 
 ss::future<> rm_stm::reset_producers() {
+    // note: must always be called under exlusive write lock to
```
nit: take units as a param?

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
/backport v24.1.x

vbotbuildovich commented:
Failed to create a backport PR to v24.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-24490-v24.2.x-345 remotes/upstream/v24.2.x
git cherry-pick -x 48371f4540 4015d4df0f 4ee6b02436
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/12277346447).

vbotbuildovich commented:
Failed to create a backport PR to v24.1.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-24490-v24.1.x-680 remotes/upstream/v24.1.x
git cherry-pick -x 48371f4540 4015d4df0f 4ee6b02436
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/12277346530).
